uro98/assignment2/PrintAncestors.java

```diff
@@ -0,0 +1,84 @@
+import java.util.Arrays;
+import java.util.LinkedList;
+
+public class PrintAncestors {
+  
+  /**
+   * Print the ancestors of the node with the given key in a binary tree as a list.
+   * Uses left-to-right depth first search.
+   * An empty list is printed if the node is not found.
+   * @param tree  a generic binary tree
+   * @param key   the target value of a node
+   */
+  public static <T> void printAncestors(BinaryTree<T> tree, T key) {
+    LinkedList<T> ancestors = new LinkedList<T>();
+    LinkedList<Node<T>> frontier = new LinkedList<Node<T>>();
+    
+    frontier.addFirst(tree.root);
+    
+    while (!frontier.isEmpty()) {
+      Node<T> node = frontier.pop();
+      
+      if (node.data.equals(key)) {
+        printList(ancestors);
+        return;
+      } else {
+        ancestors.addFirst(node.data);
+        
+        // Push the children of the node to the frontier stack, push the right child first to make DFS search left to right
+        if (node.right != null) {
+          frontier.addFirst(node.right);
+        }
+        if (node.left != null) {
+          frontier.addFirst(node.left);
+        }
+        
+        /* If the next node to be explored in the frontier is further up the tree, remove the nodes that aren't ancestors
+         * (ie the ones on the same level and below)
+         */
+        if (!frontier.isEmpty() && node.depth >= frontier.peek().depth) {
+          int depthDifference = node.depth - frontier.peek().depth;
+          for (int i = 0; i < depthDifference + 1; i++) {
+            ancestors.pop();
+          }
+        }
+      }
+    }
+    
+    // If the node is not found, print an empty list
+    ancestors.clear();
+    printList(ancestors);
+  }
+  
+  /**
+   * Print out a generic LinkedList.
+   * @param list a LinkedList
+   */
+  public static <T> void printList(LinkedList<T> list) {
+    System.out.println(Arrays.toString(list.toArray()));
+  }
+
+  public static void main(String[] args) {
+    BinaryTree<Integer> tree = new BinaryTree<Integer>();
+    
+    tree.addRoot(7);
+    
+    Node<Integer> three = new Node<Integer>(3, 1);
```

Comment:
Explicitly setting depth at the time when node is constructed and then manually setting up parent and children is not too convenient and very error-prone. It would be better if all this logic was hidden inside tree or node class.
One of possible solutions could be adding two methods in `Node`:
```
/** Adds node with given value as left child of current node and returns newly created child. */
public Node<T> addLeftChild(T value) {...}

/** Adds node with given value as left child of current node and returns newly created child. */
public Node<T> addRightChild(T value) {...}
```
Alternatively: a single `addChild` method that takes enum `Side` with values `LEFT` and `RIGHT`.

Then depth could be calculated automatically and if someone tried to use your tree in their code, they would not need to worry about manually constructing whole tree.

It would also let you make `left` and `right` fields private and provide getter methods, which would make your class much safer.